### PR TITLE
New resource: `azurerm_sentinel_watchlist_item`

### PR DESCRIPTION
**Pull request by magodo**
Add new resource `azurerm_sentinel_watchlist_item`.

The users are expected to create more than one items via uploading a csv file, where the first line represents each field, and the following lines represent the values. There is an example in the document showing how to achieve this via Terraform.

## Test

```shell
💤 TF_ACC=1 go test -timeout=3h -v ./internal/services/sentinel -run='TestAccWatchlistItem_'             
=== RUN   TestAccWatchlistItem_basic                                                                     
=== PAUSE TestAccWatchlistItem_basic                                                                     
=== RUN   TestAccWatchlistItem_update                                                                    
=== PAUSE TestAccWatchlistItem_update                                                                    
=== RUN   TestAccWatchlistItem_requiresImport                                                            
=== PAUSE TestAccWatchlistItem_requiresImport                                                            
=== CONT  TestAccWatchlistItem_basic                                                                     
=== CONT  TestAccWatchlistItem_requiresImport
=== CONT  TestAccWatchlistItem_update              
--- PASS: TestAccWatchlistItem_requiresImport (116.18s)        
--- PASS: TestAccWatchlistItem_basic (169.52s)                                                           
--- PASS: TestAccWatchlistItem_update (177.19s)                                                          
PASS                                                                                                     
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/sentinel      177.217s

```

### CODE REVIEW

File: internal/services/sentinel/sentinel_watchlist_item_resource.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+			return nil
+		},
+	}
+}
+
+func (r WatchlistItemResource) Update() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func:    r.createUpdate,
```
Comment (tombuildsstuff) [inline internal/services/sentinel/sentinel_watchlist_item_resource.go:135]: (same here) this should be a separate method, which in the case of Update retrieves and conditionally patches the resource

File: internal/services/sentinel/sentinel_watchlist_item_resource_test.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+	"github.com/hashicorp/terraform-provider-azurerm/utils"
+)
+
+type WatchlistItemResource struct {
+	Name string
+}
+
+func TestAccWatchlistItem_basic(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_sentinel_watchlist_item", "test")
+	name, _ := uuid.GenerateUUID()
```
Comment (tombuildsstuff) [inline internal/services/sentinel/sentinel_watchlist_item_resource_test.go:25]: that we're defining this in each test configuration would indicate that the resource should be generating those, and this field should be optional/computed (or even potentially removed altogether) instead?

File: website/docs/r/sentinel_watchlist_item.html.markdown [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  Manages a Sentinel Watchlist Item.
+---
+
+# azurerm_sentinel_watchlist_item
+
+Manages a Sentinel Watchlist Item.
+
+## Example Usage
+
+### Basic
```
Comment (tombuildsstuff) [inline website/docs/r/sentinel_watchlist_item.html.markdown:15]: we should only have a single example usage for most resources - more complete example usages can be placed in the examples folder?

File: website/docs/r/sentinel_watchlist_item.html.markdown [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+resource "azurerm_sentinel_watchlist_item" "example" {
+  count = length(local.csv_data)
+
+  name         = random_uuid.item[count.index].id
+  watchlist_id = azurerm_sentinel_watchlist.example.id
+  fields       = local.csv_data[count.index]
+}
+```
```
Comment (tombuildsstuff) [inline website/docs/r/sentinel_watchlist_item.html.markdown:80]: (as above) can we either remove this, or move it to the examples folder?

File: website/docs/r/sentinel_watchlist_item.html.markdown [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+* `read` - (Defaults to 5 minutes) Used when retrieving the Sentinel Watchlist Item.
+* `update` - (Defaults to 30 minutes) Used when updating the Sentinel Watchlist Item.
+* `delete` - (Defaults to 30 minutes) Used when deleting the Sentinel Watchlist Item.
+
+## Import
+
+Sentinel Watchlist Items can be imported using the `resource id`, e.g.
+
+```shell
+terraform import azurerm_sentinel_watchlist_item.example subscriptions/12345678-1234-9876-4563-123456789012/resourceGroups/resGroup1/providers/Microsoft.OperationalInsights/workspaces/workspace1/providers/Microsoft.SecurityInsights/watchlists/list1/watchlistItems/item1
```
Comment (tombuildsstuff) [inline website/docs/r/sentinel_watchlist_item.html.markdown:112]: looks like this is missing a slash?

```suggestion
terraform import azurerm_sentinel_watchlist_item.example /subscriptions/12345678-1234-9876-4563-123456789012/resourceGroups/resGroup1/providers/Microsoft.OperationalInsights/workspaces/workspace1/providers/Microsoft.SecurityInsights/watchlists/list1/watchlistItems/item1
```

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hi @magodo 

Thanks for this PR.

I've taken a look through and left some comments inline, if we can fix those up then we should be able to take another look and get this merged 👍 

Thanks!

File: internal/services/sentinel/sentinel_watchlist_item_resource.go [hunk 6]
```diff
@@ -0,0 +1,177 @@
+package sentinel
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/services/preview/securityinsight/mgmt/2019-01-01-preview/securityinsight"
```
[the remark this answers is not in the document]
Comment (magodo) [reply internal/services/sentinel/sentinel_watchlist_item_resource.go:8]: Currently, there are multiple preview versions (e.g. 2019-01-01-preview, 2021-03-01-preview), and multiple stable versions (e.g. 2020-01-01, 2021-04-01). Some resources are only included in one preview version but not the others (e.g. automation rule). Some resources are included in multiple preview versions, but not any stable version (e.g. alert rule). To make things even more complicated, some resource has less properties in newer version, e.g. there are more variants of data connectors defined in the 2019-01-01-preview than 2020-01-01.

I've offline checked with service team about above organization problem of the sentinel API. They told me that they are actually working on it and will release a version that covers all the resources. The ETA is early next year. Until that happens, we'd better stay on this version.

File: internal/services/sentinel/sentinel_watchlist_item_resource_test.go (hunk 2 above)
Comment (magodo) [reply internal/services/sentinel/sentinel_watchlist_item_resource_test.go:25]: This is the name of the watchlist item, which is a part of the URL that we have to construct.

File: website/docs/r/sentinel_watchlist_item.html.markdown (hunk 3 above)
Comment (magodo) [reply website/docs/r/sentinel_watchlist_item.html.markdown:15]: Moved to example folder.

**Comment by magodo** [comment]
@tombuildsstuff
Thanks for looking into this PR! I've resolved some of the comments while keeping the others unresolved for further discussion. Please take another review.

File: website/docs/r/sentinel_watchlist_item.html.markdown [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    k2 = "v2"
+  }
+}
+```
+
+## Arguments Reference
+
+The following arguments are supported:
+
+* `name` - (Required) The name which should be used for this Sentinel Watchlist Item. Changing this forces a new Sentinel Watchlist Item to be created.
```
Comment (katbyte) [inline website/docs/r/sentinel_watchlist_item.html.markdown:64]: if this has to be a UUID? then we should say so

and if that is the case could we just make this optional computed and automatically create one during create if name is empty?

File: website/docs/r/sentinel_watchlist_item.html.markdown [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+resource "azurerm_sentinel_watchlist" "example" {
+  name                       = "example-watchlist"
+  log_analytics_workspace_id = azurerm_log_analytics_solution.example.workspace_resource_id
+  display_name               = "example-wl"
+}
+
+resource "azurerm_sentinel_watchlist_item" "example" {
+  name         = "0aac6fa5-223e-49cf-9bfd-3554dc9d2b76"
+  watchlist_id = azurerm_sentinel_watchlist.example.id
+  fields = {
```
Comment (katbyte) [inline website/docs/r/sentinel_watchlist_item.html.markdown:53]: fields is different then many other resources? would this not be more consistent across the provider as
```suggestion
  properties = {
```

**Review by katbyte — CHANGES_REQUESTED** [review]

**Comment by magodo** [comment]
@katbyte Thank you for the new comments! I've resolved them, please take another look :)

**Comment by Andyandpandy** [comment]
Any progress on this? @katbyte @tombuildsstuff

**Comment by Zumarta** [comment]
I'd love to see this PR merged. Currently, without having the option to add items via Terraform it's hard to use it.

**Review by katbyte — APPROVED** [review]
Thanks @magodo - LGTM now

**Comment by katbyte** [comment]
once the generation check linter has been fixed this will be good to merge ⚡

**Pull request merged**